common/src/jni/main/cpp/NativeCrypto.cpp

```diff
@@ -4311,6 +4311,202 @@ static void NativeCrypto_ASN1_TIME_to_Calendar(JNIEnv* env, jclass, jlong asn1Ti
+    std::unique_ptr<CbsHandle> cbs(new CbsHandle());
+    cbs->data.reset(new unsigned char[bytes.size()]);
+    memcpy(cbs->data.get(), bytes.get(), bytes.size());
+
+    cbs->cbs.reset(new CBS());
+    CBS_init(cbs->cbs.get(), cbs->data.get(), bytes.size());
+    JNI_TRACE("asn1_read_init(%p) => %p", data, cbs.get());
+    return reinterpret_cast<uintptr_t>(cbs.release());
+}
+
+static jlong NativeCrypto_asn1_read_sequence(JNIEnv* env, jclass, jlong cbsRef) {
+    CbsHandle* cbs = reinterpret_cast<CbsHandle*>(static_cast<uintptr_t>(cbsRef));
+    JNI_TRACE("asn1_read_sequence(%p)", cbs);
+
+    std::unique_ptr<CbsHandle> seq(new CbsHandle());
+    seq->cbs.reset(new CBS());
+    if (!CBS_get_asn1(cbs->cbs.get(), seq->cbs.get(), CBS_ASN1_SEQUENCE)) {
+        Errors::throwIOException(env, "Error reading ASN.1 encoding");
+        return 0;
+    }
+    JNI_TRACE("asn1_read_sequence(%p) => %p", cbs, seq.get());
+    return reinterpret_cast<uintptr_t>(seq.release());
+}
+
+static jbyteArray NativeCrypto_asn1_read_octetstring(JNIEnv* env, jclass, jlong cbsRef) {
+    CbsHandle* cbs = reinterpret_cast<CbsHandle*>(static_cast<uintptr_t>(cbsRef));
+    JNI_TRACE("asn1_read_octetstring(%p)", cbs);
+
+    std::unique_ptr<CBS> str(new CBS());
+    if (!CBS_get_asn1(cbs->cbs.get(), str.get(), CBS_ASN1_OCTETSTRING)) {
+        Errors::throwIOException(env, "Error reading ASN.1 encoding");
+        return 0;
+    }
+    ScopedLocalRef<jbyteArray> out(env, env->NewByteArray(static_cast<jsize>(CBS_len(str.get()))));
+    if (out.get() == nullptr) {
+        Errors::throwIOException(env, "Error reading ASN.1 encoding");
+        return 0;
+    }
+    ScopedByteArrayRW outBytes(env, out.get());
+    if (outBytes.get() == nullptr) {
+        Errors::throwIOException(env, "Error reading ASN.1 encoding");
+        return 0;
+    }
+    memcpy(outBytes.get(), CBS_data(str.get()), CBS_len(str.get()));
+    JNI_TRACE("asn1_read_octetstring(%p) => %p", cbs, out.get());
+    return out.release();
+}
+
+static jlong NativeCrypto_asn1_read_integer(JNIEnv* env, jclass, jlong cbsRef) {
+    CbsHandle* cbs = reinterpret_cast<CbsHandle*>(static_cast<uintptr_t>(cbsRef));
+    JNI_TRACE("asn1_read_integer(%p)", cbs);
+
+    uint64_t value;
+    if (!CBS_get_asn1_uint64(cbs->cbs.get(), &value)) {
+        Errors::throwIOException(env, "Error reading ASN.1 encoding");
+        return 0;
+    }
+    return value;
+}
+
+static jboolean NativeCrypto_asn1_read_is_empty(CONSCRYPT_UNUSED JNIEnv* env, jclass, jlong cbsRef) {
+    CbsHandle* cbs = reinterpret_cast<CbsHandle*>(static_cast<uintptr_t>(cbsRef));
+    JNI_TRACE("asn1_read_is_empty(%p)", cbs);
+
+    bool empty = (CBS_len(cbs->cbs.get()) == 0);
+    JNI_TRACE("asn1_read_is_empty(%p) => %s", cbs, empty ? "true" : "false");
+    return empty;
+}
+
+static void NativeCrypto_asn1_read_free(CONSCRYPT_UNUSED JNIEnv* env, jclass, jlong cbsRef) {
+    if (cbsRef == 0) {
+        JNI_TRACE("asn1_read_free(0)");
+        return;
+    }
```

Comment:
Unfortunately, while delete nullptr is a no-op, reinterpret_cast<pointer type>(0) isn't guaranteed to give you the null pointer for that pointer type, so I think this circumstance does need to be special-cased.